### ISSUE TITLE
Fix compatibility of parquet.metadata with 1.5.0

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#1774]
There are quite a few incompatible changes to `parquet.metadata` (see below). It seems like it is part of the API because the metadata is exposed. Fixing it will be annoying because `ColumnPath` was removed entirely.

If this is public, then we need to fix compatibility.

```Java
Class parquet.hadoop.metadata.Canonicalizer
 Removed Class , access public super synchronized
Class parquet.hadoop.metadata.ColumnChunkMetaData
 Added Method getPath, desc ()Lparquet/common/schema/ColumnPath;, access public
 Removed Method get, sig (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;Lparquet/column/statistics/Statistics;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, desc (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;Lparquet/column/statistics/Statistics;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, access public static
 Added Method get, sig (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, desc (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, access public static
 Added Method get, sig (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;Lparquet/column/statistics/Statistics;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, desc (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;Lparquet/column/statistics/Statistics;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, access public static
 Removed Method getPath, desc ()Lparquet/hadoop/metadata/ColumnPath;, access public
 Removed Method get, sig (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, desc (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;JJJJJ)Lparquet/hadoop/metadata/ColumnChunkMetaData;, access public static
Class parquet.hadoop.metadata.ColumnChunkProperties
 Added Method getPath, desc ()Lparquet/common/schema/ColumnPath;, access public
 Removed Method getPath, desc ()Lparquet/hadoop/metadata/ColumnPath;, access public
 Removed Method get, sig (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;)Lparquet/hadoop/metadata/ColumnChunkProperties;, desc (Lparquet/hadoop/metadata/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;)Lparquet/hadoop/metadata/ColumnChunkProperties;, access public static
 Added Method get, sig (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set<Lparquet/column/Encoding;>;)Lparquet/hadoop/metadata/ColumnChunkProperties;, desc (Lparquet/common/schema/ColumnPath;Lparquet/schema/PrimitiveType$PrimitiveTypeName;Lparquet/hadoop/metadata/CompressionCodecName;Ljava/util/Set;)Lparquet/hadoop/metadata/ColumnChunkProperties;, access public static
Class parquet.hadoop.metadata.ColumnPath
 Removed Class , access final public super synchronized
```

**Reporter**: [Ryan Blue](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rdblue) / @rdblue
**Assignee**: [Ryan Blue](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rdblue) / @rdblue
#### Related issues:
- [Release parquet-mr 1.6.0](https://issues.apache.org/jira/browse/PARQUET-211) (blocks)
#### PRs and other links:
- [PR #166](https://github.com/apache/incubator-parquet-mr/pull/166)

<sub>**Note**: *This issue was originally created as [PARQUET-235](https://issues.apache.org/jira/browse/PARQUET-235). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Julien Le Dem](https://issues.apache.org/jira/browse/PARQUET-235?#comment-14482009) / @julienledem:
I don't think those were meant to be public.
possibly we want parquet.hadoop.metadata to be blacklisted in the semver plugin.

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-235?#comment-14483738) / @rdblue:
I'm currently testing a fix where parquet-common defines parquet.common.internal.ColumnPath, and both ColumnPath and Canonicalizer are added back to parquet.hadoop.metadata for compatibility. I agree that we want to limit what we're going to expose in this package after the next breaking API change, but for now I think we should go with the compatibility guarantee.

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-235?#comment-14483873) / @rdblue:
Adding PR #166 with the fix. This required moving the classes back into parquet.hadoop.metadata. The subclass strategy doesn't work because ColumnPath is final so that it can be canonicalized.

Canonicalization allows code to determine equivalence using equality, which means there can be no subclasses. The ColumnPath implementation is correctly marked final to prevent bad behavior, but it precludes options for compatible migrations.

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-235?#comment-14483965) / @rdblue:
Merged #166. Thanks for reviewing, @julienledem!
